include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -765,6 +765,36 @@ __group_scan_fits_in_slm(const sycl::queue& __queue, ::std::size_t __n, ::std::s
     return (__n <= __single_group_upper_limit && __max_slm_size >= __req_slm_size);
 }
 
+template <typename _UnaryOp>
+struct __gen_transform_input
+{
+    template <typename InRng>
+    auto
+    operator()(const InRng& __in_rng, std::size_t __idx) const
+    {
+        using _ValueType = oneapi::dpl::__internal::__value_t<InRng>;
+        using _OutValueType = oneapi::dpl::__internal::__decay_with_tuple_specialization_t<
+            typename std::invoke_result<_UnaryOp, _ValueType>::type>;
+        return _OutValueType{__unary_op(__in_rng[__idx])};
```

Comment:
As far as I can tell, all the current implementations of scan kernels use nothing like `__decay_with_tuple_specialization_t` but instead assign the result of the unary operation to either an `auto` variable or to a value of `typename _InitType::__value_type`. So I wonder what's different in this implementation? If nothing, then I would suggest to add the expected output type as the template parameter of `__gen_transform_input`, which would be set by the caller to the init value type as noted above.

Reply:
We initially ran into issues with an `auto` variable for `__in_rng[__idx]` when processing `zip_iterator` input where the dereference type is a tuple of references. It led to the caller modifying the input range when modifying the return of this function.

However, what we have here with `__decay_with_tuple_specialization_t` is unnecessary. What is important is ensuring that `__in_rng[__idx]` gets converted to the value type of the input range which can be achieved with the following:
```
        using _ValueType = oneapi::dpl::__internal::__value_t<_InRng>;
        return __unary_op(_ValueType{__in_rng[__idx]});
```
I do not think an additional template parameter is needed in this case, but we can discuss more if needed.